### DISCUSSION TITLE
Azuracast Public API in the integration of your own website

### DISCUSSION THREAD

**Discussion by radio-schlauti** [AzuraCast/AzuraCast#3576] [General]
Here are various examples of how to get information from the public API on your own website. The examples are available in PHP with the PHP tool.

Image of the moderator

```
<?php 

$station = "2"; //ID Der Station bei Azuracast
$json_station = file_get_contents("https://demo.azuracast.com/api/nowplaying/".$station);
$obj_station = json_decode($json_station);
$live = $obj_station->live->streamer_name;
$pls = $obj_station->now_playing->playlist;
?>

<img src="//Pfat zum /img/<?php if(empty($live)) {echo $pls;}else{echo $live;}
;?>.jpg" alt="<?php if(empty($live)) {echo $pls;}else{echo $live;}
?>" style="width:250; height:auto;"><br>
<?php 

$station = "2"; //ID Der Station bei Azuracast
$json_station = file_get_contents("https://demo.azuracast.com/api/nowplaying/".$station);
$obj_station = json_decode($json_station);
$live = $obj_station->live->streamer_name;
$pls = $obj_station->now_playing->playlist;
?>
```

Name of the moderator

```
<?php 

$station = "2"; //ID Der Station bei Azuracast
$json_station = file_get_contents("https://demo.azuracast.com/api/nowplaying/".$station);
$obj_station = json_decode($json_station);
$live = $obj_station->live->streamer_name;
$pls = $obj_station->now_playing->playlist;
?><?php if(empty($live)) {echo $pls;}else{echo $live;}
?>
```
Current title

```
<?php
	$station = "2";

	$json_station = file_get_contents("https://demo.azuracast.com/api/nowplaying/".$station);

	$obj_station = json_decode($json_station);

	echo "Aktuell läuft: " . $obj_station->now_playing->song->text
?>
```
Next track

```
<?php
	$station = "2";

	$json_station = file_get_contents("https://demo.azuracast.com/api/nowplaying/".$station);

	$obj_station = json_decode($json_station);

	echo "Als nächstes Kommt: " . $obj_station->playing_next->song->text	
?>
```
Important: The PHP examples do not update

maybe there is someone who is familiar with JavaScript and can expand the tools

Apache License 2.0
https://github.com/AzuraCast/AzuraCast/blob/master/LICENSE.txt

**Comment by wheslleyandrade** [comment]
Friend, I used your code and it works fine but does not update, could you help?

**Reply by PaulMaur** [reply]
It's a PHP code that can't update

**Reply by Piratenkanon** [reply]
i have created the following nowplaying.php:

[https://workupload.com/file/yshwbpczmky](url)

note: refresh every 30 seconds (line 1)

On the webpage i insert the nowplaying.php with an iframe

**Comment by honoluluman2** [comment]
For those that would like to use the API with integration to WordPress you can see below code example:

https://pastebin.com/FV5WgFiP

This should be added in functions.php in wordpress.

**Reply by ckeurk** [reply]
thank you for your code which works well, but I would like to do the same thing to display the history of the titles but I don't know how to modify your code 
https://mywebradio.com/api/station/1/history

Sample code 
https://pastebin.com/0hGS5Z6m

**Reply by honoluluman2** [reply]
The endpoint that you try to use _https://mywebradio.com/api/station/1/history_ is a secured endpoint that requires authentication which would need a modification on my original code to add security headers. A bit more complicated in general.

What you could do is just use the public endpoint https://mywebradio.com/api/nowplaying/1 that is already in my code and fetch the data from the _song_history_ property. There are data for the last 5 songs in history (depending on your azura setup).

For example, you can do:

```
$nowplaying = file_get_contents('https://demo.azuracast.com/api/nowplaying/1');
$data = json_decode( $nowplaying, true );
 
$previous_song_title_position_0 = $data['song_history']['0']['song']['title'];
$previous_song_title_position_1 = $data['song_history']['1']['song']['title'];
$previous_song_title_position_2 = $data['song_history']['2']['song']['title'];
$previous_song_title_position_3 = $data['song_history']['3']['song']['title'];
$previous_song_title_position_4 = $data['song_history']['4']['song']['title'];
```
    
and then you get the titles of the last 5 songs in history.
You should be aware that is a better practice to just use the azuracast widget that already exists for displaying the History of songs.
hope that helps

**Reply by ckeurk** [reply]
Thank you for your insight

**Reply by RadioFlamingo** [reply]
Does anyone has an script that displays the album art of the current song, and auto refresh in javascript?
I have got one that refreshes in the meta tag, but then the whole page is refreshing

**Reply by Piratenkanon** [reply]
@RadioFlamingo : simplest way is using an iframe on the page, than only the iframe refresh

**Comment by LordHelmchen666** [comment]
You can use JQuery like this example. Refresh every 15 seconds. Be sure to replace your Streamserver URL and Station name.

```
<html>
<head>
<script src="https://code.jquery.com/jquery-3.6.4.min.js" integrity="sha256-oP6HI9z1XaZNBrJURtCoUT5SUnxFr8s3BzRl+cbzUq8=" crossorigin="anonymous"></script>
</head>
<body>
<h3 class="g-title">Now Playing</h3>
<script>
var nowPlayingTimeout;
var nowPlaying;

function loadNowPlaying() {
    jQuery.ajax({
        cache: false,
        dataType: "json",
        url: 'https://STREAMSERVERURL/api/nowplaying_static/STATIONNAME.json',
        success: function(np) {
            // Do something with the Now Playing data.   
         nowPlaying = np;
		jQuery('#title').text(np.now_playing.song.title);
        jQuery('#artist').text(np.now_playing.song.artist);
    if (np.live.is_live === false ) {
            if (np.now_playing.playlist !== "") {
                jQuery('#playlist').html('<span class="badge badge-info">Playlist:  '+ np.now_playing.playlist+ '</span>');
            }  else  {
                jQuery('#playlist').html('<span class="badge badge-important"> LIVE</span>');
            }
    }  else  {
        jQuery('#playlist').html('<span class="badge badge-important">Live:  '+ np.live.streamer_name+ '</span>');
    }
        
        
            nowPlayingTimeout = setTimeout(loadNowPlaying, 15000);
        }
    }).fail(function() {
        nowPlayingTimeout = setTimeout(loadNowPlaying, 30000);
    });
}

jQuery(function() {
    loadNowPlaying();
});

</script>

<div id="title" class="g-subpromo">Titel</div>
<strong id="artist" class="g-textpromo">Artist</strong>
<div id="playlist">Sendung</div>
</body>
</html>
```

**Reply by Piratenkanon** [reply]
Hello LordHelmchen666,
Can I somehow limit the output to artist - title  ?
2 divs always create a line break
Thanks in advance

**Reply by LordHelmchen666** [reply]
Just use another tag instead of div.
`<span id="title">Titel</span> - <span id="artist">Artist</span>`
or
`<b id="title">Titel</b> - <i id="artist">Artist</i>`
or whatever you want. ;)
and delete ` <div id="playlist">Sendung</div>` if you dont wanna show playlist.

**Reply by Piratenkanon** [reply]
Thanks a lot, now i take the field #text instead of artist and title, in Text is the Format artist - title

Grüße aus Kiel

**Reply by Piratenkanon** [reply]
Hello LordHelmchen666,
Can i add also the number of listeners (listerners.total) in a line below ?
Output:
Nun Spielt: Artist - Title
Anzahl Hörer: Number of Listeners

i need this for a DJ internal page

**Reply by LordHelmchen666** [reply]
```
<html>
<head>
<script src="https://code.jquery.com/jquery-3.6.4.min.js" integrity="sha256-oP6HI9z1XaZNBrJURtCoUT5SUnxFr8s3BzRl+cbzUq8=" crossorigin="anonymous"></script>
</head>
<body>
<h3 class="g-title">Now Playing</h3>
<script>
var nowPlayingTimeout;
var nowPlaying;

function loadNowPlaying() {
    jQuery.ajax({
        cache: false,
        dataType: "json",
        url: 'https://STREAMSERVERURL/api/nowplaying_static/STATIONNAME.json',
        success: function(np) {
            // Do something with the Now Playing data.   
         nowPlaying = np;
		jQuery('#text').text('Nun spielt: '+ np.now_playing.song.text );       
                jQuery('#listeners').text('Anzahl Hörer: '+ np.listeners.total );
        
        
            nowPlayingTimeout = setTimeout(loadNowPlaying, 15000);
        }
    }).fail(function() {
        nowPlayingTimeout = setTimeout(loadNowPlaying, 30000);
    });
}

jQuery(function() {
    loadNowPlaying();
});

</script>

<div id="text">Titel</div>
<div id="listeners">Hörer</div>
</body>
</html>
```

**Comment by Moonbase59** [comment]
Wouldn’t it be much more elegant to just create a new web hook in AzuraCast and have it _push_ the data to your website? _Pulling is evil._ ;-)

This is a snap, since AzuraCast nicely provides us with custom web hooks, that will fire automatically whenever the now playing data changes. Plus, it’ll give you all that JSON data in the body.

Nice examples though, as always @LordHelmchen666!

**Reply by BusterNeece** [reply]
It's a little harder to write the relevant code to accept an incoming push from our webhooks, write that out to a temp cache file, then server that temp cache file out to listeners. If you directly use our Centrifugo-based high-performance updates in your client-side JS, then you get the full benefit of the push-based NowPlaying updates with relatively little implementation work, but even that's more complex than just stock jQuery calls.

**Comment by Moonbase59** [comment]
Just for completeness, if anyone is still interested in this: SSE (Server-sent events) client in this [Gist](https://gist.github.com/Moonbase59/d42f411e10aff6dc58694699010307aa), with [Instructions](https://gist.github.com/Moonbase59/d42f411e10aff6dc58694699010307aa?permalink_comment_id=4786603#gistcomment-4786603).

Due to the many changes in the high-performance now-playing lately, this works with the current rolling releases (and hopefully future stable releases) but _not_ with old stable releases (0.19.3 and older).

**Comment by unudelaradio** [comment]
Hello!

Could someone please help me with a jquery code like @LordHelmchen666 examples, but for Artwork/AlbumArt?

**Reply by LordHelmchen666** [reply]
I recommend SSE script from Moonebase59. It pushes data only on song changes and don't poll every x seconds.
> : SSE (Server-sent events) client in this [Gist](https://gist.github.com/Moonbase59/d42f411e10aff6dc58694699010307aa), with [Instructions](https://gist.github.com/Moonbase59/d42f411e10aff6dc58694699010307aa?permalink_comment_id=4786603#gistcomment-4786603).
> 

If really jQuery version needed
```

<html>
<head>
<script src="https://code.jquery.com/jquery-3.7.1.min.js" integrity="sha256-/JqT3SQfawRcv/BIHPThkBvs0OEvtFFmqPF/lYI/Cxo=" crossorigin="anonymous"></script>
</head>
<body>
<h3 class="g-title">Now Playing</h3>
<script>
var nowPlayingTimeout;
var nowPlaying;

function loadNowPlaying() {
    jQuery.ajax({
        cache: false,
        dataType: "json",
        url: 'https://STREAMURL/api/nowplaying_static/STATION.json',
        success: function(np) {
            // Do something with the Now Playing data.   
         nowPlaying = np;
		jQuery('#albumart').html('<img src=" '+ np.now_playing.song.art +' " width=100 >');       
        
        
            nowPlayingTimeout = setTimeout(loadNowPlaying, 15000);
        }
    }).fail(function() {
        nowPlayingTimeout = setTimeout(loadNowPlaying, 30000);
    });
}

jQuery(function() {
    loadNowPlaying();
});

</script>

<div id="albumart">Albumart</div>
</body>
</html>
```

**Reply by unudelaradio** [reply]
Lots of thanks!

I tested today SSE implementation, but for some reasons, don't know why, on my website works fine only for one station (a station with non stop streaming). When i test the HTML file on my computer, everything is ok, everytime i change shortcodes, no matter the station. But when i insert the scripts to website, loading is very lazy... or for some stations doesn't show anything at all. Another strange behaviour is that on Firefox works, but on Brave and Edge is not working... must be something with my website. But jQuery is loading instantly.
